tensorflow/python/framework/python_op_gen_test.cc

```diff
@@ -20,22 +20,269 @@ limitations under the License.
+    name: "y"
+    type_attr: "T2"
+  }
+  output_arg {
+    name: "output"
+    type_attr: "T"
+  }
+  attr {
+    name: "T"
+    type: "type"
+    allowed_values {
+      list {
+        type: DT_UINT8
+        type: DT_INT8
+      }
+    }
+  }
+  attr {
+    name: "T2"
+    type: "type"
+    allowed_values {
+      list {
+        type: DT_STRING
+        type: DT_FLOAT
+        type: DT_DOUBLE
+      }
+    }
+  }
+  summary: "Summary for op Foo."
+  description: "Description for op Foo."
+}
+op {
+  name: "Bar"
+  input_arg {
+    name: "x"
+    type: DT_STRING
+  }
+  input_arg {
+    name: "y"
+    type: DT_QINT8
+  }
+  output_arg {
+    name: "output"
+    type: DT_BOOL
+  }
+  summary: "Summary for op Bar."
+  description: "Description for op Bar."
+}
+op {
+  name: "FooBar"
+  input_arg {
+    name: "x"
+    type: DT_FLOAT
+  }
+  output_arg {
+    name: "output"
+    type: DT_BOOL
+  }
+  attr {
+    name: "t"
+    type: "type"
+    allowed_values {
+      list {
+        type: DT_HALF
+        type: DT_INT8
+      }
+    }
+  }
+  attr {
+    name: "var1"
+    type: "bool"
+    default_value {
+      b: false
+    }
+  }
+  attr {
+    name: "var2"
+    type: "int"
+    default_value {
+      i: 0
+    }
+  }
+  summary: "Summary for op FooBar."
+  description: "Description for op FooBar."
+}
+op {
+  name: "Baz"
+  input_arg {
+    name: "inputs"
+    number_attr: "N"
+    type_list_attr: "T"
+  }
+  output_arg {
+    name: "output1"
+    type: DT_BOOL
+  }
+  output_arg {
+    name: "output2"
+    type: DT_BOOL
+  }
+  attr {
+    name: "T"
+    type: "bool"
+  }
+  attr {
+    name: "N"
+    type: "int"
+  }
+  summary: "Summary for op Baz."
+  description: "Description for op Baz."
+}
+)";
+
+std::unordered_set<string> type_annotate_ops {
+  "Foo",
+  "Bar",
+  "FooBar",
+  "Baz"
+};
+
+
+void ExpectHasSubstr(const string& s, const string& expected) {
+  EXPECT_TRUE(absl::StrContains(s, expected))
+      << "'Generated ops does not contain '" << expected << "'";
+}
+
+void ExpectDoesNotHaveSubstr(const string& s, const string& expected) {
+  EXPECT_FALSE(absl::StrContains(s, expected))
+      << "'Generated ops contains '" << expected << "'";
+}
+
+void ExpectSubstrOrder(const string& s, const string& before,
+                       const string& after) {
+  int before_pos = s.find(before);
+  int after_pos = s.find(after);
+  ASSERT_NE(std::string::npos, before_pos);
+  ASSERT_NE(std::string::npos, after_pos);
+  EXPECT_LT(before_pos, after_pos)
+      << before << "' is not before '" << after;
+}
+
 TEST(PythonOpGen, Basic) {
   OpList ops;
   OpRegistry::Global()->Export(false, &ops);
 
   ApiDefMap api_def_map(ops);
 
-  string code = GetPythonOps(ops, api_def_map, {}, "");
+  string code = GetPythonOps(ops, api_def_map, {}, "", {});
 
   EXPECT_TRUE(absl::StrContains(code, "def case"));
-
   // TODO(mdan): Add tests to verify type annotations are correctly added.
 }
 
+TEST(PythonOpGen, TypeAnnotateSingleTypeTensor) {
```

Comment:
These tests are good. Consider declaring the proto string inside the function function itself, so that one can read everything there is to know about the test without scrolling back and forth.

Reply:
This sounds good to me, I put the proto strings inside each test in `b00c93e`.